src/IO/H5/Helpers.cpp

```diff
@@ -303,8 +304,22 @@ Type read_value_attribute(const hid_t location_id, const std::string& name) {
   const hid_t attribute_id = H5Aopen(location_id, name.c_str(), h5p_default());
   CHECK_H5(attribute_id, "Failed to open attribute '" << name << "'");
   Type value;
-  CHECK_H5(H5Aread(attribute_id, h5_type<Type>(), &value),
-           "Failed to read attribute '" << name << "'");
+  if constexpr (std::is_same_v<Type, std::string>) {
+    // The HDF5 c API returns a c string, which must be converted to a
+    // std::string. c string must have a fixed size; here choose a size
+    // much larger than a conceivable input file could be, but not so large
+    // as to have a sizable impact on memory usage. The largest yaml file
+    // in tests/InputFiles as of this writing is ~8k characters.
+    // Here, set the maximum string size to 10000000 = 10M characters, which
+    // occupies 10MB of memory.
+    char* value_c = new char[10000000];
```

Comment:
HDF5 limits this to 64K.  That's not obviously large enough for future input files.

Reply:
Did not know about that limit for HDF5. Thanks for pointing it out, @wthrowe!

I think there are good reasons to put the yaml file in an attribute, even with this limit. It's easy for users to access with whatever tools they like (h5dump prints it nicely, users reading it in from python need not worry about converting to string), and functionally the yaml file is metadata, not data.

Here are the sizes of all of our input files so far. The largest is the BBH input file, which is still 8x smaller than the 64k limit. I would argue that if input files are approaching the limit, then probably there's too much specified in them, instead of in code.

```
2058 ./Elasticity/BentBeam.yaml
    2857 ./Elasticity/HalfSpaceMirror.yaml
    2383 ./ScalarWave/PlaneWave1DObserveExample.yaml
    2028 ./ScalarWave/PlaneWave1DEventsAndTriggersExample.yaml
    2094 ./ScalarWave/PlaneWave2D.yaml
    1435 ./ScalarWave/PlaneWave3D.yaml
    1757 ./ScalarWave/PlaneWave1D.yaml
    1402 ./CurvedScalarWave/PlaneWaveMinkowski2D.yaml
    1414 ./CurvedScalarWave/PlaneWaveMinkowski1D.yaml
    2939 ./CurvedScalarWave/PlaneWaveMinkowski3D.yaml
    1670 ./Burgers/Step.yaml
    3994 ./Xcts/KerrSchild.yaml
    5335 ./Xcts/BinaryBlackHole.yaml
    1658 ./RelativisticEuler/Valencia/SmoothFlow2D.yaml
    1690 ./RelativisticEuler/Valencia/SmoothFlow1D.yaml
    1682 ./RelativisticEuler/Valencia/SmoothFlow3D.yaml
    1581 ./RadiationTransport/M1Grey/ConstantM1.yaml
    1328 ./Cce/AnalyticTestTeukolskyWave.yaml
    1361 ./Cce/AnalyticTestLinearizedBondiSachs.yaml
    1345 ./Cce/AnalyticTestBouncingBlackHole.yaml
    1381 ./Cce/AnalyticTestGaugeWave.yaml
    1548 ./Cce/AnalyticTestRobinsonTrautman.yaml
    1330 ./Cce/AnalyticTestRotatingSchwarzschild.yaml
    1517 ./Cce/CharacteristicExtract.yaml
    3315 ./GrMhd/ValenciaDivClean/FishboneMoncriefDisk.yaml
    2401 ./GrMhd/ValenciaDivClean/BlastWave.yaml
    5142 ./GrMhd/GhValenciaDivClean/GhMhdBondiMichel.yaml
    4461 ./GrMhd/GhValenciaDivClean/GhMhdTovStar.yaml
    4551 ./GeneralizedHarmonic/KerrSchild.yaml
    2832 ./GeneralizedHarmonic/GaugeWave3D.yaml
    2920 ./GeneralizedHarmonic/GaugeWave1D.yaml
    9669 ./GeneralizedHarmonic/BinaryBlackHole.yaml
    1980 ./Poisson/ProductOfSinusoids3D.yaml
    2002 ./Poisson/ProductOfSinusoids1D.yaml
    1940 ./Poisson/ProductOfSinusoids2D.yaml
    1503 ./NewtonianEuler/RiemannProblem2D.yaml
    1537 ./NewtonianEuler/RiemannProblem1D.yaml
    1526 ./NewtonianEuler/RiemannProblem3D.yaml
     939 ./ExportCoordinates/Input2D.yaml
     964 ./ExportCoordinates/Input3D.yaml
     904 ./ExportCoordinates/Input1D.yaml
    2202 ./ExportCoordinates/InputTimeDependent3D.yaml
     149 ./ExampleExecutables/SingletonHelloWorld.yaml
     124 ./ExampleExecutables/HelloWorldNoCharm.yaml
    1627 ./ScalarAdvection/Krivodonova1D.yaml
    1612 ./ScalarAdvection/Sinusoid1D.yaml
    1566 ./ScalarAdvection/Kuzmin2D.yaml
```

So I prefer, after thinking about it, using an attribute to store the source input (though I'm open to arguments otherwise).

The alternative, if I understand it, is to store this as data (i.e. as an array of integer values). Is there a straightforward way to output ints to data in H5 files in spectre, or would this functionality have to be added? When I looked, I only saw functionality for outputting doubles, aside from the source tarball.

Reply:
HDF5 does have some support for "large attributes": http://davis.lbl.gov/Manuals/HDF5-1.8.7/UG/13_Attributes.html#SpecIssues .  As far as I know we don't have any support for string-like output.

Also, I just noticed the `write_to_attribute` function immediately following this.  Can you use that here?

Reply:
Sorry, could you please clarify your question? Here I'm reading the attribute, not writing it, so I'm confused what you're asking me to do.

Thanks for the very quick review, by the way!

Reply:
Oops.  Sorry.  Ignore that.

Reply:
I just tested (using h5py, python 3.7, and hdf5 1.12.1) that I could write an attribute of 1e7 random alphanumeric characters, using the script below. I noticed no noticable performance issues, the output file was 10MB as expected, and h5dump could display the attribute no problem. So in practice, I don't think input file size is likely to be a concern when storing the input file as an HDF5 attribute.

~~~
import h5py
import random
import string

stringLength=10000000
bigAttr = ''.join(random.choices(string.ascii_letters + string.digits, k=stringLength))
bigAttr[-4:]

f = h5py.File('testBigAttr.h5', 'w')
f.attrs.create("InputSource.yaml", bigAttr)
f.close()

f2 = h5py.File('testBigAttr.h5', 'r')
readBigAttr = f2.attrs["InputSource.yaml"]
print(readBigAttr == bigAttr)
f2.close()
~~~

Reply:
Weird.  The h5py developers claim that doesn't work: https://github.com/h5py/h5py/issues/1855#issuecomment-795804097

Can you test writing with the C++ code, in case h5py is doing something different from us?  Adding a giant comment to the input file should be easy enough.